src/main/java/com/conveyal/datatools/manager/jobs/feedmerge/MergeLineContext.java

```diff
@@ -237,42 +237,104 @@ public void startNewRow() throws IOException {
             .collect(Collectors.toList());
     }
 
+    public enum ReferenceTableLookup {
+        TRIP_SERVICE_ID_KEY(
+            String.join(":", Table.TRIPS.name, SERVICE_ID, Table.CALENDAR.name, Table.CALENDAR_DATES.name)
+        );
+
+        private final String value;
+
+        ReferenceTableLookup(String value) {
+            this.value = value;
+        }
+
+        public String getValue() {
+            return value;
+        }
+
+        public static ReferenceTableLookup fromValue(String key) {
+            for (ReferenceTableLookup ref: ReferenceTableLookup.values()) {
+                if (ref.getValue().equals(key)) {
+                    return ref;
+                }
+            }
+            throw new UnsupportedOperationException(String.format("Unsupported reference table combination: %s.", key));
+        }
+    }
+
+    /**
+     * Determine which reference table to use. If there is only one reference use this. If there are multiple references
+     * determine the context and then the correct reference table to use.
+     */
+    private Table getReferenceTable(FieldContext fieldContext, Field field) {
+        if (field.referenceTables.size() == 1) {
+            return field.referenceTables.iterator().next();
+        }
+
+        // A table reference is still required even if no match is found.
+        Table defaultTable = null;
+        switch (ReferenceTableLookup.fromValue(createKey(field))) {
+            case TRIP_SERVICE_ID_KEY:
+                boolean isCalendarServiceId = mergeFeedsResult.calendarServiceIds.contains(fieldContext.getValueToWrite());
+                boolean isCalendarDatesServiceId = mergeFeedsResult.calendarDatesServiceIds.contains(fieldContext.getValueToWrite());
+                if (isCalendarServiceId && !isCalendarDatesServiceId) {
+                    return Table.CALENDAR;
+                } else if (!isCalendarServiceId && isCalendarDatesServiceId) {
+                    return Table.CALENDAR_DATES;
+                } else {
+                    // For this case it doesn't seem to matter which is returned so going with calendar.
+                    defaultTable = Table.CALENDAR;
+                }
+            // Add other cases as multiple references are added e.g. flex!.
+        }
+        return defaultTable;
+    }
+
+    /**
+     * Create a unique key for this table, field and reference tables.
+     */
+    private String createKey(Field field) {
+        return String.format(
+            "%s:%s:%s",
+            table.name,
+            field.name,
+            field.referenceTables.stream().map(r -> r.name).collect(Collectors.joining(":"))
```

Comment:
Hopefully we don't rely extract values from keys by relying on the ":" separator.

Reply:
I've made changes to this in my latest update to use a more unque key.